### PR TITLE
Gains `fake_factset_financial_data`

### PR DESCRIPTION
**Pull request by jdhoffa**
A function that, by default, outputs a 1-row fake version of `factset_financial_data`.

Relates to #4

### CODE REVIEW

File: NAMESPACE [hunk 1]
```diff
@@ -1,2 +1,4 @@
 # Generated by roxygen2: do not edit by hand
 
+export(fake_factset_financial_data)
+importFrom(rlang,"%||%")
```
Comment (cjyetman) [inline NAMESPACE:4]: nit: I struggle to remember precisely what this custom infix operator does, and therefore prefer not to use it in code that I write. {rlang} is also a rather heavy dependency just for a single operator like this.

File: R/fake_factset_financial_data.R [hunk 2]
```diff
@@ -0,0 +1,35 @@
+#' Minimal fake `factset_financial_data` dataset that allow overwriting values
+#'
+#' @template fake_data
+#'
+#' @return A data.frame
```
Comment (cjyetman) [inline R/fake_factset_financial_data.R:5]: strictly speaking, a `tibble` (or `tbl_df`?), no?

File: R/fake_factset_financial_data.R [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+                                        isin = NULL,
+                                        factset_entity_id = NULL,
+                                        adj_price = NULL,
+                                        adj_shares_outstanding = NULL,
+                                        issue_type = NULL,
+                                        one_adr_eq = NULL,
+                                        ...) {
+  #TODO: Add checks that the input arguments match expectations, and error if not
+  # * check that inputs are the correct type
+  # * check that inputs are the expected length and char/numeric combination (e.g. for ISINs)
```
Comment (cjyetman) [inline R/fake_factset_financial_data.R:22]: comment: while my initial thought was "yes, please!" especially for ISINs, then I considered a potential desired use case of creating a fake `factset_financial_data` object that intentionally has an invalid ISIN, which might imply that something like a `strict = TRUE` option might be useful eventually

File: R/fake_factset_financial_data.R [hunk 4]
```diff
@@ -0,0 +1,35 @@
+#' Minimal fake `factset_financial_data` dataset that allow overwriting values
+#'
+#' @template fake_data
```
Comment (cjyetman) [inline R/fake_factset_financial_data.R:3]: does this import parameter descriptions from elsewhere? otherwise, that's required, no?

File: R/fake_factset_financial_data.R [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+    fsym_id = fsym_id %||% default_values[["fsym_id"]],
+    isin = isin %||% default_values[["isin"]],
+    factset_entity_id =
+      factset_entity_id %||% default_values[["factset_entity_id"]],
+    adj_price = adj_price %||% default_values[["adj_price"]],
+    adj_shares_outstanding =
+      adj_shares_outstanding %||% default_values[["adj_shares_outstanding"]],
+    issue_type = issue_type %||% default_values[["issue_type"]],
+    one_adr_eq = one_adr_eq %||% default_values[["one_adr_eq"]],
+    ...
```
Comment (cjyetman) [inline R/fake_factset_financial_data.R:33]: nit: how this `...` will behave and/or what are valid inputs for it is hard for me to reason about without experimentation

File: R/imports.R [hunk 6]
```diff
@@ -0,0 +1,2 @@
+#' @importFrom rlang %||%
```
Comment (cjyetman) [inline R/imports.R:1]: nit: I think this is a {usethis} style of importing? I prefer explicitly setting imports in the Roxygen comments above the function that uses them, which makes it easier to manage imports, especially e.g. if you remove a function that needs an import and it's the only function in the package that was using it

File: R/utils.R [hunk 7]
```diff
@@ -0,0 +1,10 @@
+default_values <- list(
+  fsym_id = "A12345-B",
+  isin = "AB123CD4E567",
+  factset_entity_id = "1234A5-B",
+  adj_price = 12.34567,
+  adj_shares_outstanding = 1234567890,
+  issue_type = "EQ",
+  # one_adr_eq typically has values of 1, 0.25, 0.33, and other perfect fractions
+  one_adr_eq = 1/4
+)
```
Comment (cjyetman) [inline R/utils.R:10]: nit: if I'm being really strict about hyper-focussed PRs, I would prefer to see this added in a separate PR

File: R/utils.R [hunk 8]
```diff
@@ -0,0 +1,10 @@
+default_values <- list(
+  fsym_id = "A12345-B",
+  isin = "AB123CD4E567",
+  factset_entity_id = "1234A5-B",
```
Comment (cjyetman) [inline R/utils.R:4]: nit: there are probably better, more realistic yet still fake examples for these, but I'm not currently in a position to make suggestions

File: tests/testthat/test-fake_factset_financial_data.R [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+  names <- names(out)
+  expected_names <- c(
+    "fsym_id",
+    "isin",
+    "factset_entity_id",
+    "adj_price",
+    "adj_shares_outstanding",
+    "issue_type",
+    "one_adr_eq"
+  )
```
Comment (cjyetman) [inline tests/testthat/test-fake_factset_financial_data.R:13]: should these not be taken directly from the `default_values` object? otherwise if they ever change, they need to be changed here too?

File: tests/testthat/test-fake_factset_financial_data.R [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+
+  expect_identical(sort(names), sort(expected_names))
+
+})
+
+test_that("by default, has expected values", {
```
Comment (cjyetman) [inline tests/testthat/test-fake_factset_financial_data.R:19]: probably should also test that it has 1-row, since that's a very clear intention?

File: R/fake_factset_financial_data.R (hunk 3 above)
Comment (jdhoffa) [reply R/fake_factset_financial_data.R:22]: Yeah, as you'll see in a future push I thought the exact same thing and updated the comment XD

One option would be to warn when it is invalid? That requires us to be sure we are doing it on purpose.

File: R/fake_factset_financial_data.R (hunk 4 above)
Comment (jdhoffa) [reply R/fake_factset_financial_data.R:3]: Yes, I have drafted a bunch of "template" documentation in the `man-roxygen` folder. 
I figure every `fake_*` function will have 90% of the same documentation, so modularized it.

File: R/fake_factset_financial_data.R (hunk 5 above)
Comment (jdhoffa) [reply R/fake_factset_financial_data.R:33]: Functions that same as in `tibble::tibble()`.
So, if you call `fake_factset_financial_data(foo = "bar")` it will output a tibble, with a new column `foo` with the value `baz`. 

* if the tibble output has > 1 rows, it will copy `bar` to every row
* if the tibble output has > 1 rows and the input to `foo` is something like `c("bar", "baz") it will work only if the output tibble has exactly 2 rows, otherwise it will fail with an informative error

File: R/imports.R (hunk 6 above)
Comment (jdhoffa) [reply R/imports.R]: Removed, and dropped `rlang` dep as requested

File: R/fake_factset_financial_data.R (hunk 3 above)
Comment (cjyetman) [reply R/fake_factset_financial_data.R:22]: I guess that's a good reason for this repo to be currently "experimental", because honestly I don't have a strong opinion about it, yet 😉

File: R/utils.R (hunk 7 above)
Comment (jdhoffa) [reply R/utils.R:10]: This is required by the `fake_factset_financial_data` function, otherwise, it will fail.

File: R/utils.R (hunk 8 above)
Comment (jdhoffa) [reply R/utils.R:4]: For sure, my thought process was the following:
* I wanted the default value to represent generally an expected value (same type, same length, same broad order of characters, strings, etc)
* but, I also wanted it to be abundantly, obviously, fundamentally clear to anyone seeing the data that it was fake, and doesn't correspond to anything real

I think those two qualities are important to preserve, but happy to discuss it further.

File: tests/testthat/test-fake_factset_financial_data.R (hunk 9 above)
Comment (jdhoffa) [reply tests/testthat/test-fake_factset_financial_data.R:13]: I don't think so... here is why:
The point of `default_values` is that it is NOT specific to `factset_financial_data`. 
Every future `fake_*` data function will take the `default_value` by default (e.g. for `fsym_id`), and in general the `default_values` list should grow to accommodate every possible column name. 

(I recognize this may get a bit crazy, and we may want a `factset_default_values` or something like that)

but the general point is that, while it is true now, I don't expect it to always be true that the names of `default_values` aligns 1-to-1 with the names of `financial_data`

Comment (cjyetman) [reply tests/testthat/test-fake_factset_financial_data.R:13]: 👍🏻 but then just for my clarity.... if a FactSet column value changed names, one would need to edit the `default_value` specification plus every test instance where it was explicitly used?

Comment (jdhoffa) [reply tests/testthat/test-fake_factset_financial_data.R:13]: Not necessarily... 
In the `fake_*` function, the line looks something like:
``` r
#...
isin = isin %||% default_values[["isin"]]
```

if you changed the name of the column in the dataset, it MAY just then look like:
``` r
new_isin_name = new_isin_name %||% default_values[["isin"]]
```

... though not sure if that would be strategic or desirable. Let's explore how it works in the wild as we add more datasets? I am not committed to any of this, just prototyping things and testing it out.

File: R/utils.R [hunk 11]
```diff
@@ -0,0 +1,16 @@
+default_values <- list(
+  fsym_id = "A12345-B",
```
Comment (AlexAxthelm) [inline R/utils.R:2]: style nit: for lists like this, order doesn't matter, but it makes the dev process easier if they're sorted (fewer merge conflicts, easier to find what you're looking for)

Comment (jdhoffa) [reply R/utils.R:2]: Yeah that makes sense!

File: NAMESPACE (hunk 1 above)
Comment (jdhoffa) [reply NAMESPACE]: I have removed the dependency on `rlang`, and defined the "infix operator" explicitly in the package (you're right, it's a very bulky dependency for such a small function).

Have a look at the function trace of `%||%` in the package, it is pretty simple but very useful functionality (especially in the context where default arguments to a function are `NULL`)

NIT: I do typically use `rlang` very frequently, however, so it may be coming back at some point

File: R/fake_factset_financial_data.R (hunk 2 above)
Comment (jdhoffa) [reply R/fake_factset_financial_data.R]: Fixed.

File: R/fake_factset_financial_data.R (hunk 3 above)
Comment (jdhoffa) [reply R/fake_factset_financial_data.R]: See #7

File: R/fake_factset_financial_data.R (hunk 5 above)
Comment (jdhoffa) [reply R/fake_factset_financial_data.R:49]: Added better documentation for that parameter. If you feel the documentation should be more verbose, please feel free to suggest changes!

File: tests/testthat/test-fake_factset_financial_data.R (hunk 9 above)
Comment (jdhoffa) [reply tests/testthat/test-fake_factset_financial_data.R:21]: NB/NIT/Thinking aloud:
I believe the value names in `default_values` should always be the minimally descriptive name for the column/ data type (e.g. `isin`), even if the column name in some dataset ends up being `factset_equity_share_cool_value_isin`)

File: tests/testthat/test-fake_factset_financial_data.R (hunk 10 above)
Comment (jdhoffa) [reply tests/testthat/test-fake_factset_financial_data.R]: Tested

File: R/utils.R (hunk 11 above)
Comment (jdhoffa) [reply R/utils.R]: Done

**Comment by jdhoffa** [comment]
Ok, I have:
* Addressed all comments
* Added more (and more precise) testing

I believe this is ready for review :-)

File: R/fake_factset_financial_data.R [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+    if_arg_exists_check_valid_type(fsym_id, is.character),
+    if_arg_exists_check_valid_type(isin, is.character),
+    if_arg_exists_check_valid_type(factset_entity_id, is.character),
+    if_arg_exists_check_valid_type(adj_price, is.numeric),
+    if_arg_exists_check_valid_type(adj_shares_outstanding, is.numeric),
+    if_arg_exists_check_valid_type(issue_type, is.character),
+    if_arg_exists_check_valid_type(one_adr_eq, is.numeric)
+  )
+
+    tibble::tibble(
```
Comment (cjyetman) [inline R/fake_factset_financial_data.R:39]:
```suggestion
  tibble::tibble(
```

File: tests/testthat/test-fake_factset_financial_data.R [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+
+  expect_identical(sort(names), sort(expected_names))
+
+})
+
+test_that("default output has expected values", {
+  out <- fake_factset_financial_data()
+
+  for (name in names(out)) {
+    expect_equal(out[[!!name]], default_values[[!!name]])
```
Comment (cjyetman) [inline tests/testthat/test-fake_factset_financial_data.R:73]: what is the `!!` for here?

**Review by cjyetman — DISMISSED** [review]

File: R/fake_factset_financial_data.R [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+    adj_price = adj_price %||% default_values[["adj_price"]],
+    adj_shares_outstanding =
+      adj_shares_outstanding %||% default_values[["adj_shares_outstanding"]],
+    issue_type = issue_type %||% default_values[["issue_type"]],
+    one_adr_eq = one_adr_eq %||% default_values[["one_adr_eq"]],
+    ...
+  )
+}
+
+if_arg_exists_check_valid_type <- function(arg, fun) {
```
Comment (AlexAxthelm) [inline R/fake_factset_financial_data.R:53]: NB: Suggest moving this into `utils.R`

File: R/utils.R [hunk 15]
```diff
@@ -0,0 +1,14 @@
+default_values <- list(
```
Comment (AlexAxthelm) [inline R/utils.R:1]: NB: This feels important enough to split into its own file.

File: tests/testthat/test-fake_factset_financial_data.R [hunk 16]
```diff
@@ -0,0 +1,97 @@
+test_that("default output is 1-row 7-col tibble", {
+  out <- fake_factset_financial_data()
+  expect_equal(nrow(out), 1L)
+  expect_equal(ncol(out), 7L)
+  expect_true(inherits(out, "tbl_out") || inherits(out, "tbl"))
```
Comment (AlexAxthelm) [inline tests/testthat/test-fake_factset_financial_data.R:5]: NB (question): Why is the OR condition needed here?

File: tests/testthat/test-fake_factset_financial_data.R [hunk 17]
```diff
@@ -0,0 +1,97 @@
...
+  out <- fake_factset_financial_data()
+
+  for (name in names(out)) {
+    expect_equal(out[[!!name]], default_values[[!!name]])
+  }
+})
+
+test_that("with bad data, errors informatively", {
+
+  args <- c(
```
Comment (AlexAxthelm) [inline tests/testthat/test-fake_factset_financial_data.R:79]: NB: Why is this necessary? it doesn't seem to be called.

**Review by AlexAxthelm — APPROVED** [review]
No blocking comments.

File: tests/testthat/test-fake_factset_financial_data.R (hunk 13 above)
Comment (jdhoffa) [reply tests/testthat/test-fake_factset_financial_data.R:73]: Just developer-facing improvements. The `!!` here gives more informative messages if the test fails, so that future @jdhoffa has an easier time debugging.

Compare this (with no `!!`):
Screenshot 2024-02-16 at 08 54 58

to this (with `!!):
Screenshot 2024-02-16 at 08 55 29

Comment (cjyetman) [reply tests/testthat/test-fake_factset_financial_data.R:73]: interesting... so `!!` is the {rlang} "injection" thing?

File: R/fake_factset_financial_data.R (hunk 14 above)
Comment (jdhoffa) [reply R/fake_factset_financial_data.R]: Agreed. I went a step further and made it an internal dataset, with minimal documentation in a code comment. At the moment, I don't think we should export it, but would be happy to discuss that point later.

File: R/utils.R (hunk 15 above)
Comment (jdhoffa) [reply R/utils.R]: Agreed. I went a step further and made it an internal dataset, with minimal documentation in a code comment. At the moment, I don't think we should export it, but would be happy to discuss that point later.

File: tests/testthat/test-fake_factset_financial_data.R (hunk 16 above)
Comment (jdhoffa) [reply tests/testthat/test-fake_factset_financial_data.R]: Honestly just copied that in from a different package. 
Made it more explicit.

File: tests/testthat/test-fake_factset_financial_data.R (hunk 17 above)
Comment (jdhoffa) [reply tests/testthat/test-fake_factset_financial_data.R]: stray code

File: tests/testthat/test-fake_factset_financial_data.R (hunk 13 above)
Comment (AlexAxthelm) [reply tests/testthat/test-fake_factset_financial_data.R:73]: Neat.

**Comment by cjyetman** [comment]
FYI... I just re-ran the windows-latest (4.1) test because it looks like it just hiccuped on the {pak} install

File: tests/testthat/test-fake_factset_financial_data.R (hunk 13 above)
Comment (jdhoffa) [reply tests/testthat/test-fake_factset_financial_data.R:73]: Yup the "injection operator" or "bang bang" (lol...) injects the value or expression into another expression.
It is very useful for programming and testing with `dplyr` (or packages that use data-masking more generally)

**Review by AlexAxthelm — DISMISSED** [review]
🚀

File: DESCRIPTION [hunk 18]
```diff
@@ -12,3 +12,8 @@ Config/testthat/edition: 3
 Encoding: UTF-8
 Roxygen: list(markdown = TRUE)
 RoxygenNote: 7.3.1
+Imports: 
+    tibble
+Depends: 
+    R (>= 2.10)
```
Comment (cjyetman) [inline DESCRIPTION:18]: ummmm.... are sure you want to support that? LOL

**Review by cjyetman — APPROVED** [review]

File: DESCRIPTION (hunk 18 above)
Comment (jdhoffa) [reply DESCRIPTION:18]: Lol it was automatically added when I added Lazy data. 
But it seems that `tibble` requires `3.4.0`, so I will add that instead :-)

**Review by cjyetman — APPROVED** [review]

**Review by AlexAxthelm — APPROVED** [review]
